### PR TITLE
add CI - build & run [wip]

### PR DESCRIPTION
**Pull request by Borda**
adding CI which build the code and execute sample run

### CODE REVIEW

**Comment by Borda** [comment]
@sarthakpati I am trying to add CI which build the binary and try to run the example, I am following the steps in the README but it seems to be failing on Linux, mind have look on the failer? https://github.com/Borda/HistoReg/actions

**Comment by sarthakpati** [comment]
> @sarthakpati I am trying to add CI which build the binary and try to run the example, I am following the steps in the README but it seems to be failing on Linux, mind have look on the failer? https://github.com/Borda/HistoReg/actions

Looking at it now.

File: .github/workflows/ci-build-run.yml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        mkdir bin
+
+    - name: build for Linux
+      if: runner.os == 'linux'
+      run: |
+        cd bin
+        # this will build ITK as HistoReg needs
+        cmake ..
+        # add '-j${N}' to specify number of threads to use for compilation
+        make -j${N}
```
Comment (sarthakpati) [inline .github/workflows/ci-build-run.yml:34]: This should be `make`

File: .github/workflows/ci-build-run.yml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        ## wait for ITK to finish building
+        # this is for HistoReg
+        cmake -DCMAKE_INSTALL_PREFIX=$pwd/"ITK-build" ..
+        # add '-j${N}' to specify number of threads to use for compilation
+        make -j${N}
```
Comment (sarthakpati) [inline .github/workflows/ci-build-run.yml:39]: This should be `make`

File: .github/workflows/ci-build-run.yml [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      if: runner.os == 'linux'
+      run: |
+        cd bin
+        # this will build ITK as HistoReg needs
+        cmake ..
+        # add '-j${N}' to specify number of threads to use for compilation
+        make -j${N}
+        ## wait for ITK to finish building
+        # this is for HistoReg
+        cmake -DCMAKE_INSTALL_PREFIX=$pwd/"ITK-build" ..
```
Comment (sarthakpati) [inline .github/workflows/ci-build-run.yml:37]: This should be `cmake -DITK_DIR=./ITK-build -DCMAKE_INSTALL_PREFIX=./install ..`

File: .github/workflows/ci-build-run.yml [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    - name: build for Windows
+      if: runner.os == 'windows'
+      run: |
+        cd bin
+        # this will build ITK as HistoReg needs
+        cmake ..
+        cmake --build . --target ALL_BUILD
+        ## wait for ITK to finish building
+        # this is for HistoReg
+        cmake -DCMAKE_INSTALL_PREFIX=$pwd\"ITK-build" ..
```
Comment (sarthakpati) [inline .github/workflows/ci-build-run.yml:52]: This should be `cmake -DITK_DIR=./ITK-build -DCMAKE_INSTALL_PREFIX=./install ..`

**Comment by sarthakpati** [comment]
So, I had put in the instructions assuming someone would want to do multi-threaded compiles. I believe that is why the build is failing. I am not very familiar with using GH actions (I mostly use Azure DevOps), so I cannot be certain, though.

Does this help?

File: .github/workflows/ci-build-run.yml [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  pytest:
+
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ubuntu-18.04, windows-2019]
+
+    # Timeout: https://stackoverflow.com/a/59076067/4521646
+    timeout-minutes: 50
```
Comment (sarthakpati) [inline .github/workflows/ci-build-run.yml:17]: Depending on the hardware, ITK compiles might take longer, unfortunately. 😞

**Comment by Borda** [comment]
> > @sarthakpati I am trying to add CI which build the binary and try to run the example, I am following the steps in the README but it seems to be failing on Linux, mind have look on the failer? https://github.com/Borda/HistoReg/actions
> 
> Looking at it now.

cool, thank you! mind use suggestions in your comments si I can later simply accept them
(eventually, you can edit your comments and make them as suggestions, see [Suggesting Changes on GitHub](https://haacked.com/archive/2019/06/03/suggested-changes/))

File: .github/workflows/ci-build-run.yml (hunk 5 above)
Comment (sarthakpati) [inline .github/workflows/ci-build-run.yml:17]:
```suggestion
    timeout-minutes: 200
```

File: .github/workflows/ci-build-run.yml (hunk 1 above)
Comment (sarthakpati) [inline .github/workflows/ci-build-run.yml:34]:
```suggestion
        make 
```

File: .github/workflows/ci-build-run.yml (hunk 3 above)
Comment (sarthakpati) [inline .github/workflows/ci-build-run.yml:37]:
```suggestion
        cmake -DITK_DIR=./ITK-build -DCMAKE_INSTALL_PREFIX=./install ..
```

File: .github/workflows/ci-build-run.yml (hunk 2 above)
Comment (sarthakpati) [inline .github/workflows/ci-build-run.yml:39]:
```suggestion
        make 
```

File: .github/workflows/ci-build-run.yml [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    - name: build for Windows
+      if: runner.os == 'windows'
+      run: |
+        cd bin
+        # this will build ITK as HistoReg needs
+        cmake ..
+        cmake --build . --target ALL_BUILD
+        ## wait for ITK to finish building
+        # this is for HistoReg
+        cmake -DITK_DIR=./ITK-build -DCMAKE_INSTALL_PREFIX=./install ..`
```
Comment (sarthakpati) [inline .github/workflows/ci-build-run.yml:52]:
```suggestion
        cmake -DITK_DIR=./ITK-build -DCMAKE_INSTALL_PREFIX=./install ..
```

File: .github/workflows/ci-build-run.yml [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        cmake --build . --target ALL_BUILD
+        cmake --build . --target INSTALL
+
+    - name: sample Run
+      run: |
+        HistoReg_source=${pwd}
+        HistoReg_install=${pwd}
```
Comment (sarthakpati) [inline .github/workflows/ci-build-run.yml:59]:
```suggestion
        HistoReg_install=${pwd}/bin/install
```

File: .github/workflows/ci-build-run.yml [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        ${HistoReg_install}/bin/HistoReg -m ${HistoReg_source}/Data/Images/CD68.jpg -f ${HistoReg_source}/Data/Images/CD4.jpg -c /path/to/c2d/executable  -o ${HistoReg_source}/Data/Output/ -l ${HistoReg_source}/Data/Landmarks/CD68.csv -S
```
Comment (sarthakpati) [inline .github/workflows/ci-build-run.yml:60]:
```suggestion
        ${HistoReg_install}/bin/HistoReg -m ${HistoReg_source}/Data/Images/CD68.jpg -f ${HistoReg_source}/Data/Images/CD4.jpg -o ${HistoReg_source}/Data/Output/ -l ${HistoReg_source}/Data/Landmarks/CD68.csv -S
```

**Comment by sarthakpati** [comment]
Unsure why it wasn't able to find the install tree; I am testing on [a separate branch](https://github.com/sarthakpati/HistoReg/pull/2/checks?check_run_id=1058067470), if you don't mind.

**Comment by Borda** [comment]
Well maybe we can change the destination brach to yours and merge it... Then you can continue debugging?

**Comment by sarthakpati** [comment]
> Well maybe we can change the destination brach to yours and merge it... Then you can continue debugging?

Sure, works for me.

**Comment by sarthakpati** [comment]
Closing because CI has been added via Azure Pipelines (GH actions kept for Linux).

**Pull request closed without merging**